File: daily/2026/3742_Maximum_Path_Score_in_a_Grid.cpp
#include <iostream>
#include <vector>

using namespace std;

class Solution {
public:
    int maxPathScore(vector<vector<int>>& grid, int k) {
        int m = grid.size();
        int n = grid[0].size();
        vector<vector<vector<int>>> dp(m, vector<vector<int>>(n, vector<int>(k + 2, -1)));
        int c = (grid[0][0] > 0)? 1 : 0;
        int s = grid[0][0];
        dp[0][0][c] = s;
        for (int j = 1; j < n; j++){
            if (grid[0][j] > 0) c++, s += grid[0][j];
            if (c > k) break;
            dp[0][j][c] = s;
        }
        c = (grid[0][0] > 0)? 1 : 0;
        s = grid[0][0];
        for (int i = 1; i < m; i++){
            if (grid[i][0] > 0) c++, s += grid[i][0];
            if (c > k) break;
            dp[i][0][c] = s;
        }
        for (int i = 1; i < m; i++){
            for (int j = 1; j < n; j++){
                int curr = grid[i][j];
                if (curr == 0) dp[i][j][0] = max(dp[i-1][j][0], dp[i][j-1][0]);
                for (int t = 1; t <= k; t++){
                    if (dp[i-1][j][t - (curr > 0)] != -1) dp[i][j][t] = max(dp[i][j][t], dp[i-1][j][t - (curr > 0)] + curr);
                    if (dp[i][j-1][t - (curr > 0)] != -1) dp[i][j][t] = max(dp[i][j][t], dp[i][j-1][t - (curr > 0)] + curr);
                    // dp[i][j][t] = max(dp[i-1][j][t - (curr > 0)], dp[i][j-1][t - (curr > 0)]) + curr;
                }
            }
        }
        int ans = -1;
        for (int t = 0; t <= k; t++) ans = max(ans, dp[m-1][n-1][t]);
        return ans;
    }
};

int main(){
    vector<vector<int>> grid = {{0,1},{2,0}};
    int k = 1;
    // vector<vector<int>> grid = {{0,1},{1,2}};
    // int k = 1;
    // vector<vector<int>> grid = {{0, 2, 2, 2, 2, 2, 2, 2, 2, 2}};
    // int k = 5;
    Solution S;
    cout << S.maxPathScore(grid, k) << endl;
    return 0;
}

File: daily_python/3742_Maximum_Path_Score_in_a_Grid.py
import os
import sys
from typing import List

class Solution:
    def maxPathScore(self, grid: List[List[int]], k: int) -> int:
        m, n = len(grid), len(grid[0])
        dp = [[[-1] * (k + 2) for _ in range(n)] for _ in range(m)]
        if grid[0][0] == 0:
            dp[0][0][0] = 0
        else:
            dp[0][0][1] = grid[0][0]
        for i in range(m):
            for j in range(n):
                for c in range(k + 1):
                    if dp[i][j][c] == -1:
                        continue
                    if j + 1 < n:
                        val = grid[i][j+1]
                        cost = 1 if val > 0 else 0
                        dp[i][j+1][c + cost] = max(dp[i][j+1][c + cost], dp[i][j][c] + val)
                    if i + 1 < m:
                        val = grid[i+1][j]
                        cost = 1 if val > 0 else 0
                        dp[i+1][j][c + cost] = max(dp[i+1][j][c + cost], dp[i][j][c] + val)
        ans = -1
        for c in range(k + 1):
            ans = max(ans, dp[m-1][n-1][c])
        return ans

def readDataSet(filename):
    dataset = []
    with open(filename, 'r') as file:
        content = file.read().strip()
        blocks = content.split('\n\n')
        for block in blocks:
            lines = block.split('\n')
            grid = [list(map(int, row.split(','))) for row in lines[0].split('=')[1].strip()[2:-3].split('},{')]
            k = int(lines[1].split('=')[1].strip()[:-1])
            dataset.append((grid, k))
    return dataset

if __name__ == '__main__':
    if len(sys.argv) == 1:
        filename = os.path.splitext(__file__)[0] + '.txt'
    else:
        filename = sys.argv[1]
    dataset = readDataSet(filename)
    results = []
    solution = Solution()
    for grid, k in dataset:
        results.append(solution.maxPathScore(grid, k))
    for index, result in enumerate(results):
        print(f'Example {index + 1} : {result}')